ResearchKit/Common/ORKDiscreteGraphView.m

```diff
@@ -0,0 +1,1023 @@
+            }
+            
+            if (![positionOnYAxis isRangeZero]) {
+                
+                CGFloat pointSize = self.isLandscapeMode ? 10.0f : 8.0f;
+                ORKCircleView *point = [[ORKCircleView alloc] initWithFrame:CGRectMake(0, 0, pointSize, pointSize)];
+                point.tintColor = (plotIndex == 0) ? self.tintColor : self.referenceLineColor;
+                point.center = CGPointMake(positionOnXAxis, positionOnYAxis.maximumValue);
+                [self.plotsView.layer addSublayer:point.layer];
+                
+                if (self.shouldAnimate) {
+                    point.alpha = 0;
+                }
+                
+                [self.dots addObject:point];
+            }
+        }
+    }
+}
+
+- (void)drawLinesForPlotIndex:(NSInteger)plotIndex {
+    CGFloat positionOnXAxis = CGFLOAT_MAX;
+    ORKRangePoint *positionOnYAxis = nil;
+    
+    for (NSUInteger i=0; i<self.yAxisPoints.count; i++) {
+        
+        ORKRangePoint *dataPointVal = self.dataPoints[i];
+        
+        if (!dataPointVal.isEmpty && !dataPointVal.isRangeZero) {
+            
+            UIBezierPath *plotLinePath = [UIBezierPath bezierPath];
+            
+            positionOnXAxis = [self.xAxisPoints[i] floatValue];
+            positionOnXAxis += [self offsetForPlotIndex:plotIndex];
+                
+            positionOnYAxis = ((ORKRangePoint *)self.yAxisPoints[i]);
+            
+            [plotLinePath moveToPoint:CGPointMake(positionOnXAxis, positionOnYAxis.minimumValue)];
+            
+            [plotLinePath addLineToPoint:CGPointMake(positionOnXAxis, positionOnYAxis.maximumValue)];
+            
+            CAShapeLayer *plotLineLayer = [CAShapeLayer layer];
+            plotLineLayer.path = plotLinePath.CGPath;
+            plotLineLayer.fillColor = [UIColor clearColor].CGColor;
+            plotLineLayer.strokeColor = (plotIndex == 0) ? self.tintColor.CGColor : self.referenceLineColor.CGColor;
+            plotLineLayer.lineJoin = kCALineJoinRound;
+            plotLineLayer.lineCap = kCALineCapRound;
+            plotLineLayer.lineWidth = self.isLandscapeMode ? 10.0 : 8.0;
+            plotLineLayer.opacity = 0.4;
+            [self.plotsView.layer addSublayer:plotLineLayer];
+            
+            if (self.shouldAnimate) {
+                plotLineLayer.strokeEnd = 0;
+            }
+            [self.pathLines addObject:plotLineLayer];
+            
+        }
+    }
+}
+
+- (CGFloat)offsetForPlotIndex:(NSInteger)plotIndex {
+    CGFloat pointWidth = self.isLandscapeMode ? 10.0 : 8.0;
+    
+    NSInteger numberOfPlots = [self numberOfPlots];
+    
+    CGFloat offset = 0;
+    
+    if (numberOfPlots%2 == 0) {
+        //Even
+        offset = (plotIndex - numberOfPlots/2 + 0.5) * pointWidth;
+    } else {
+        //Odd
+        offset = (plotIndex - numberOfPlots/2) * pointWidth;
+    }
+    
+    return offset;
+}
+
+#pragma mark - Graph Calculations
```

Comment:
Improve code reuse (after using the same *data source protocol*) from here to the end of the file,